schema/query/projection_evaluator.go

```diff
@@ -39,25 +39,63 @@ func (p Projection) Eval(ctx context.Context, payload map[string]interface{}, rs
 	return payload, err
 }
 
-func evalProjection(ctx context.Context, p Projection, payload map[string]interface{}, validator schema.Validator, rbr *referenceBatchResolver) (map[string]interface{}, error) {
-	res := map[string]interface{}{}
-	resMu := sync.Mutex{} // XXX use sync.Map once Go 1.9 is out
+func prepareProjection(p Projection, payload map[string]interface{}) (Projection, error) {
+	var proj Projection
 	if len(p) == 0 {
 		// When the Projection is empty, it's like saying "all fields".
 		// This allows notations like id,user{} to embed all fields of the user
 		// sub-resource.
 		for fn := range payload {
-			p = append(p, ProjectionField{Name: fn})
+			proj = append(proj, ProjectionField{Name: fn})
+		}
+		return proj, nil
+	}
+
+	hasStar := false
+	var starChildren Projection
+	for _, pf := range p {
+		if pf.Name == "*" {
+			if hasStar {
+				return nil, fmt.Errorf("only one * in projection allowed")
+			}
+			hasStar = true
+			starChildren = pf.Children
+		} else {
+			proj = append(proj, pf)
+		}
+	}
+	if hasStar {
+		for fn := range payload {
+			exists := false
+			for _, pf := range proj {
+				if fn == pf.Name && pf.Alias == "" {
+					exists = true
+				}
+			}
+			if !exists {
+				proj = append(proj, ProjectionField{Name: fn, Children: starChildren})
+			}
 		}
 	}
+	return proj, nil
+}
+
+func evalProjection(ctx context.Context, p Projection, payload map[string]interface{}, fg schema.FieldGetter, rbr *referenceBatchResolver) (map[string]interface{}, error) {
+	res := map[string]interface{}{}
+	resMu := sync.Mutex{} // XXX use sync.Map once Go 1.9 is out
```

Comment:
Cam remove FIXME. SyncMap also probably isn't fit for purpose here.

Reply:
Removed in separate commit.

Reply:
@smyrman Did you noticed that I have changed method parameter from `validator schema.Validator` to `fg schema.FieldGetter`?